File: nsfw/src/main/java/com/zwy/nsfw/api/NsfwHelper.java
package com.zwy.nsfw.api;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.graphics.Bitmap;
import android.util.Log;
import com.zwy.nsfw.Classifier;

import java.io.IOException;

public class NsfwHelper {
    @SuppressLint("StaticFieldLeak")
    private static NsfwHelper nsfwHelper;
    private Activity activity;
    private Classifier classifier;

    public static NsfwHelper getInstance(Activity activity, Boolean isAddGpuDelegate, int numThreads) {
        synchronized (NsfwHelper.class) {
            if (nsfwHelper == null) {
                nsfwHelper = new NsfwHelper(activity, isAddGpuDelegate, numThreads);
            }
        }
        return nsfwHelper;
    }

    private NsfwHelper(Activity activity, Boolean isAddGpuDelegate, int numThreads) {
        try {
            this.activity = activity;
            classifier = Classifier.create(activity, isAddGpuDelegate, numThreads);
        } catch (IOException e) {
            e.printStackTrace();
            Log.e(Classifier.TAG, "Tensorflow Lite Image Classifier Initialization Error,e:" + e);
        }
    }


    /**
     * 同步扫描
     *
     * @param bitmap
     * @return nsfw/sfw
     */
    public NsfwBean scanBitmapSyn(Bitmap bitmap) {
        synchronized (NsfwHelper.class) {
            return classifier.run(bitmap);
        }
    }


    /**
     * 异步扫描
     *
     * @param bitmap
     * @param onScanBitmapListener void onSuccess(float sfw, float nsfw);
     */
    public void scanBitmap(final Bitmap bitmap, final OnScanBitmapListener onScanBitmapListener) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                final NsfwBean nsfwBean = scanBitmapSyn(bitmap);
                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        onScanBitmapListener.onSuccess(nsfwBean.getSfw(), nsfwBean.getNsfw());
                    }
                });
            }
        }).start();
    }

    public void destroy() {
        classifier.close();
        activity = null;
        nsfwHelper = null;
    }

    public static interface OnScanBitmapListener {
        void onSuccess(float sfw, float nsfw);
    }


}

File: nsfw/src/main/java/com/zwy/nsfw/api/NSFWHelper.kt
package com.zwy.nsfw.api

import android.graphics.Bitmap
import com.zwy.nsfw.core.Classifier
import com.zwy.nsfw.core.NSFWConfig
import com.zwy.nsfw.core.NsfwBean

object NSFWHelper {

    private lateinit var classifier: Classifier

    private var isInit = false

    @Synchronized
    fun init(nsfwConfig: NSFWConfig): NSFWHelper {
        classifier = Classifier.create(nsfwConfig.assetMannager, nsfwConfig.userGPU, 10)
        isInit = true
        return this
    }


    /**
     * 同步扫描图片(建议使用者自己放在线程中处理)
     */
    @Synchronized
    fun scanBitmap(bitmap: Bitmap): NsfwBean {
        if (!isInit) {
            throw ExceptionInInitializerError("Initialize the scanned object 'NSFWHelper.init(nsfwConfig: NSFWConfig)' by calling the function first.")
        }
        return classifier.run(bitmap)
    }


    /**
     * 不会销毁该单利和配置的参数，建议页面处理完成后调用该代码对扫描器进行关闭,关闭后下次扫描不需要再次调用init方法
     */
    fun destroyFactory() {
        classifier.close()
    }


}